File: validate_intrv_complete.f90
program validate_intrv_complete
    use intrv_module
    use, intrinsic :: iso_fortran_env, only: real32
    implicit none
    
    ! Maximum array size and number of tests
    integer, parameter :: max_array_size = 100
    integer, parameter :: max_tests = 500
    
    ! Test data storage
    real :: xt_array(max_array_size)
    integer :: lxt, ilo_in, ilo_out
    real :: x_value
    integer :: ileft, mflag
    
    ! Expected values
    integer :: expected_ileft, expected_mflag, expected_ilo
    
    ! Counters and statistics
    integer :: test_id, i, array_size
    integer :: total_tests = 0
    integer :: passed_tests = 0
    integer :: failed_tests = 0
    
    ! Failure tracking
    integer :: extreme_value_failures = 0
    integer :: boundary_failures = 0
    integer :: binary_search_failures = 0
    integer :: mflag_failures = 0
    
    ! File handling
    integer :: unit_in = 10
    integer :: ios
    character(len=1000) :: line
    character(len=100) :: dummy
    
    ! Open results file
    open(unit=20, file='intrv_validation_results.txt', status='replace')
    write(20,'(A)') 'INTRV Modern Implementation Validation Report'
    write(20,'(A)') '============================================='
    write(20,'(A)') ''
    
    ! Read test data
    open(unit=unit_in, file='test_data/intrv_tests.json', status='old', iostat=ios)
    if (ios /= 0) then
        write(*,*) 'Error: Cannot open test_data/intrv_tests.json'
        stop 1
    end if
    
    ! Skip header lines until we find test_cases
    do
        read(unit_in, '(A)', iostat=ios) line
        if (ios /= 0) exit
        if (index(line, '"test_cases"') > 0) exit
    end do
    
    ! Process each test case
    do test_id = 1, max_tests
        ! Skip to next test
        do
            read(unit_in, '(A)', iostat=ios) line
            if (ios /= 0) exit
            if (index(line, '"test_id":') > 0) then
                ! Found a test, now parse it
                call parse_test_case()
                exit
            end if
        end do
        
        if (ios /= 0) exit
        
        ! Run the test
        ilo_out = ilo_in
        call intrv(xt_array, lxt, x_value, ilo_out, ileft, mflag)
        
        total_tests = total_tests + 1
        
        ! Check results
        if (ileft == expected_ileft .and. mflag == expected_mflag .and. &
            ilo_out == expected_ilo) then
            passed_tests = passed_tests + 1
        else
            failed_tests = failed_tests + 1
            
            ! Analyze failure type
            call analyze_failure()
            
            ! Report failure details for extreme value tests
            if (is_extreme_value_test()) then
                write(20,'(A,I3,A)') 'Test ', test_id, ' (Extreme Value Test) FAILED:'
                write(20,'(A,ES12.5)') '  X value: ', x_value
                write(20,'(A)') '  Array values:'
                do i = 1, lxt
                    write(20,'(A,I2,A,ES12.5)') '    XT(', i, ') = ', xt_array(i)
                end do
                write(20,'(A,I3,A,I3,A,I3)') '  Expected: ILEFT=', expected_ileft, &
                    ', MFLAG=', expected_mflag, ', ILO=', expected_ilo
                write(20,'(A,I3,A,I3,A,I3)') '  Got:      ILEFT=', ileft, &
                    ', MFLAG=', mflag, ', ILO=', ilo_out
                write(20,'(A)') ''
            end if
        end if
    end do
    
    close(unit_in)
    
    ! Write summary
    write(20,'(A)') ''
    write(20,'(A)') 'Summary Statistics:'
    write(20,'(A)') '==================='
    write(20,'(A,I4)') 'Total tests:     ', total_tests
    write(20,'(A,I4)') 'Passed tests:    ', passed_tests
    write(20,'(A,I4)') 'Failed tests:    ', failed_tests
    write(20,'(A,F6.2,A)') 'Pass rate:       ', &
        real(passed_tests) / real(total_tests) * 100.0, '%'
    write(20,'(A)') ''
    write(20,'(A)') 'Failure Analysis:'
    write(20,'(A,I4)') 'Extreme value failures:    ', extreme_value_failures
    write(20,'(A,I4)') 'Boundary condition failures:', boundary_failures
    write(20,'(A,I4)') 'Binary search failures:    ', binary_search_failures
    write(20,'(A,I4)') 'MFLAG failures:            ', mflag_failures
    
    close(20)
    
    ! Console output
    write(*,'(A)') 'INTRV validation complete.'
    write(*,'(A,I4,A,I4,A,F6.2,A)') 'Results: ', passed_tests, '/', total_tests, &
        ' passed (', real(passed_tests) / real(total_tests) * 100.0, '%)'
    write(*,'(A)') 'Detailed results written to intrv_validation_results.txt'
    
contains

    subroutine parse_test_case()
        ! Parse array values
        read(unit_in, '(A)') line  ! skip "inputs" line
        read(unit_in, '(A)') line  ! skip "[" line
        
        ! Read array values
        array_size = 0
        do
            read(unit_in, '(A)') line
            if (index(line, ']') > 0) exit
            array_size = array_size + 1
            read(line, *) xt_array(array_size)
        end do
        
        ! Read LXT
        read(unit_in, '(A)') line
        read(line, *) lxt
        
        ! Read X value
        read(unit_in, '(A)') line
        read(line, *) x_value
        
        ! Read initial ILO
        read(unit_in, '(A)') line
        read(line, *) ilo_in
        
        ! Skip to expected values
        do
            read(unit_in, '(A)') line
            if (index(line, '"expected"') > 0) exit
        end do
        
        ! Read expected ILEFT
        read(unit_in, '(A)') line
        read(line(index(line,':')+1:), *) expected_ileft
        
        ! Read expected MFLAG
        read(unit_in, '(A)') line
        read(line(index(line,':')+1:), *) expected_mflag
        
        ! Read expected ILO_out
        read(unit_in, '(A)') line
        read(line(index(line,':')+1:), *) expected_ilo
    end subroutine
    
    logical function is_extreme_value_test()
        integer :: i
        logical :: has_extreme
        
        has_extreme = .false.
        
        ! Check for very small values
        do i = 1, lxt
            if (abs(xt_array(i)) < 1.0e-20 .and. xt_array(i) /= 0.0) then
                has_extreme = .true.
                exit
            end if
        end do
        
        ! Check for very large values
        do i = 1, lxt
            if (abs(xt_array(i)) > 1.0e20) then
                has_extreme = .true.
                exit
            end if
        end do
        
        ! Check if x is extreme
        if (abs(x_value) < 1.0e-20 .and. x_value /= 0.0) has_extreme = .true.
        if (abs(x_value) > 1.0e20) has_extreme = .true.
        
        is_extreme_value_test = has_extreme
    end function
    
    subroutine analyze_failure()
        ! Categorize the failure
        
        ! Check if it's an extreme value issue
        if (is_extreme_value_test()) then
            extreme_value_failures = extreme_value_failures + 1
        end if
        
        ! Check if it's a boundary condition
        if (x_value < xt_array(1) .or. x_value >= xt_array(lxt)) then
            boundary_failures = boundary_failures + 1
        end if
        
        ! Check if MFLAG is wrong
        if (mflag /= expected_mflag) then
            mflag_failures = mflag_failures + 1
        end if
        
        ! Otherwise it's likely a binary search issue
        if (.not. is_extreme_value_test() .and. &
            x_value >= xt_array(1) .and. x_value < xt_array(lxt)) then
            binary_search_failures = binary_search_failures + 1
        end if
    end subroutine

end program validate_intrv_complete

File: validate_intrv_complete.py
#!/usr/bin/env python3
"""
Validate INTRV modern implementation against F77 reference values
"""

import json
import subprocess
import tempfile
import os
from typing import List, Tuple, Dict

def create_test_program(xt_array: List[float], lxt: int, x: float, ilo_in: int) -> str:
    """Create a Fortran test program for a single test case"""
    
    # Format array values with proper precision
    array_str = ", ".join(f"{val:e}" for val in xt_array[:lxt])
    
    program = f"""
program test_intrv_single
    use intrv_module
    implicit none
    
    real :: xt({lxt}) = (/ {array_str} /)
    integer :: lxt = {lxt}
    real :: x = {x:e}
    integer :: ilo = {ilo_in}
    integer :: ileft, mflag
    
    call intrv(xt, lxt, x, ilo, ileft, mflag)
    
    ! Output results
    write(*,'(I0,",",I0,",",I0)') ileft, mflag, ilo
    
end program test_intrv_single
"""
    return program

def run_single_test(xt_array: List[float], lxt: int, x: float, ilo_in: int) -> Tuple[int, int, int]:
    """Run a single INTRV test and return (ileft, mflag, ilo_out)"""
    
    # Create temporary test program
    with tempfile.NamedTemporaryFile(mode='w', suffix='.f90', delete=False) as f:
        f.write(create_test_program(xt_array, lxt, x, ilo_in))
        temp_filename = f.name
    
    try:
        # Compile
        exec_name = temp_filename.replace('.f90', '')
        compile_result = subprocess.run(
            ['gfortran', '-o', exec_name, 'modern/intrv_modern.f90', temp_filename],
            capture_output=True, text=True
        )
        
        if compile_result.returncode != 0:
            print(f"Compilation error: {compile_result.stderr}")
            return None, None, None
        
        # Run
        run_result = subprocess.run([exec_name], capture_output=True, text=True)
        
        if run_result.returncode != 0:
            print(f"Runtime error: {run_result.stderr}")
            return None, None, None
        
        # Parse output
        output = run_result.stdout.strip()
        ileft, mflag, ilo_out = map(int, output.split(','))
        
        return ileft, mflag, ilo_out
        
    finally:
        # Cleanup
        os.unlink(temp_filename)
        if os.path.exists(exec_name):
            os.unlink(exec_name)

def is_extreme_value_test(xt_array: List[float], x: float) -> bool:
    """Check if this test involves extreme floating-point values"""
    extreme_small = 1e-20
    extreme_large = 1e20
    
    # Check array values
    for val in xt_array:
        if val != 0.0 and (abs(val) < extreme_small or abs(val) > extreme_large):
            return True
    
    # Check x value
    if x != 0.0 and (abs(x) < extreme_small or abs(x) > extreme_large):
        return True
    
    return False

def main():
    # Load test data
    with open('test_data/intrv_tests.json', 'r') as f:
        data = json.load(f)
    
    total_tests = len(data['test_cases'])
    passed_tests = 0
    failed_tests = 0
    
    # Failure categorization
    failures_by_category = {
        'extreme_value': [],
        'boundary': [],
        'binary_search': [],
        'mflag': []
    }
    
    # Special tests to watch
    special_tests = [254, 255, 257]
    
    print("INTRV Modern Implementation Validation")
    print("=====================================")
    print(f"Running {total_tests} tests...\n")
    
    # Run each test
    for i, test_case in enumerate(data['test_cases']):
        test_id = test_case['test_id']
        xt_array = test_case['inputs'][0]
        lxt = test_case['inputs'][1]
        x = test_case['inputs'][2]
        ilo_in = test_case['inputs'][3]
        
        expected_ileft = test_case['expected']['ileft']
        expected_mflag = test_case['expected']['mflag']
        expected_ilo = test_case['expected']['ilo_out']
        
        # Run test
        ileft, mflag, ilo_out = run_single_test(xt_array, lxt, x, ilo_in)
        
        if ileft is None:
            print(f"Test {test_id}: ERROR - Could not run test")
            failed_tests += 1
            continue
        
        # Check results
        if ileft == expected_ileft and mflag == expected_mflag and ilo_out == expected_ilo:
            passed_tests += 1
            if test_id in special_tests:
                print(f"Test {test_id}: PASSED (Special test)")
        else:
            failed_tests += 1
            
            # Categorize failure
            if is_extreme_value_test(xt_array[:lxt], x):
                failures_by_category['extreme_value'].append(test_id)
            
            if x < xt_array[0] or x >= xt_array[lxt-1]:
                failures_by_category['boundary'].append(test_id)
            
            if mflag != expected_mflag:
                failures_by_category['mflag'].append(test_id)
            
            if (x >= xt_array[0] and x < xt_array[lxt-1] and 
                not is_extreme_value_test(xt_array[:lxt], x)):
                failures_by_category['binary_search'].append(test_id)
            
            # Report special test failures
            if test_id in special_tests or is_extreme_value_test(xt_array[:lxt], x):
                print(f"\nTest {test_id}: FAILED")
                print(f"  Array: {xt_array[:lxt]}")
                print(f"  X value: {x}")
                print(f"  Initial ILO: {ilo_in}")
                print(f"  Expected: ILEFT={expected_ileft}, MFLAG={expected_mflag}, ILO={expected_ilo}")
                print(f"  Got:      ILEFT={ileft}, MFLAG={mflag}, ILO={ilo_out}")
                
                # Analyze the issue
                if ileft == 4 and mflag == 0 and ilo_out == 4:
                    print("  Issue: Returned suspicious pattern (4, 0, 4)")
        
        # Progress indicator
        if (i + 1) % 50 == 0:
            print(f"Progress: {i + 1}/{total_tests} tests completed...")
    
    # Write detailed report
    with open('intrv_validation_report.txt', 'w') as report:
        report.write("INTRV Modern Implementation Validation Report\n")
        report.write("=============================================\n\n")
        
        report.write(f"Summary Statistics:\n")
        report.write(f"==================\n")
        report.write(f"Total tests:     {total_tests}\n")
        report.write(f"Passed tests:    {passed_tests}\n")
        report.write(f"Failed tests:    {failed_tests}\n")
        report.write(f"Pass rate:       {passed_tests/total_tests*100:.2f}%\n\n")
        
        report.write(f"Failure Analysis:\n")
        report.write(f"=================\n")
        report.write(f"Extreme value failures:     {len(failures_by_category['extreme_value'])}\n")
        report.write(f"Boundary condition failures: {len(failures_by_category['boundary'])}\n")
        report.write(f"Binary search failures:     {len(failures_by_category['binary_search'])}\n")
        report.write(f"MFLAG failures:             {len(failures_by_category['mflag'])}\n\n")
        
        # Detailed failure analysis
        if failures_by_category['extreme_value']:
            report.write("Extreme Value Test Failures:\n")
            report.write("---------------------------\n")
            for test_id in failures_by_category['extreme_value'][:10]:  # First 10
                test = data['test_cases'][test_id-1]
                report.write(f"Test {test_id}: {test['description']}\n")
            if len(failures_by_category['extreme_value']) > 10:
                report.write(f"... and {len(failures_by_category['extreme_value'])-10} more\n")
            report.write("\n")
    
    # Console summary
    print(f"\n\nValidation Complete!")
    print(f"===================")
    print(f"Total tests:     {total_tests}")
    print(f"Passed tests:    {passed_tests}")
    print(f"Failed tests:    {failed_tests}")
    print(f"Pass rate:       {passed_tests/total_tests*100:.2f}%")
    print(f"\nDetailed report written to intrv_validation_report.txt")

if __name__ == "__main__":
    main()